flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogHandleReaderWithCache.java

```diff
@@ -0,0 +1,184 @@
+            cacheFileCreator;
+    private final ConcurrentMap<Path, RefCountedBufferingFileStream> cache =
+            new ConcurrentHashMap<>();
+    private final ScheduledExecutorService cacheCleanScheduler;
+    private final long cacheIdleMillis;
+
+    ChangelogHandleReaderWithCache(Configuration config) {
+        File[] tempFiles =
+                Arrays.stream(ConfigurationUtils.parseTempDirectories(config))
+                        .map(path -> new File(path, CACHE_FILE_SUB_DIR))
+                        .toArray(File[]::new);
+
+        this.cacheCleanScheduler =
+                SchedulerFactory.create(1, "ChangelogCacheFileCleanScheduler", LOG);
+        // TODO: 2022/5/31 consider adding a new options for cache idle
+        this.cacheIdleMillis = config.get(PERIODIC_MATERIALIZATION_INTERVAL).toMillis();
+        this.cacheFileCreator = RefCountedTmpFileCreator.inDirectories(tempFiles);
+    }
+
+    @Override
+    public DataInputStream apply(StreamStateHandle handle, Long offset) throws IOException {
+        if (!(handle instanceof FileStateHandle)) {
+            return wrapAndSeek(handle.openInputStream(), offset);
+        }
+
+        FileStateHandle fileHandle = (FileStateHandle) handle;
+        DataInputStream input;
+
+        if (fileHandle.getFilePath().getFileSystem().isDistributedFS()) {
+
+            Path dfsPath = fileHandle.getFilePath();
+
+            final RefCountedBufferingFileStream refCountedFileStream =
+                    cache.computeIfAbsent(
+                            dfsPath,
+                            key -> {
+                                RefCountedBufferingFileStream fileStream = null;
+                                FSDataInputStream handleInputStream = null;
+
+                                try {
+                                    fileStream =
+                                            RefCountedBufferingFileStream.openNew(cacheFileCreator);
+                                    handleInputStream = handle.openInputStream();
+                                    IOUtils.copyBytes(wrap(handleInputStream), fileStream);
+                                    LOG.debug(
+                                            "download and decompress dstl file : {} to cached file : {}",
+                                            fileHandle.getFilePath(),
+                                            fileStream.getInputFile().getPath());
+                                } catch (IOException e) {
+                                    ExceptionUtils.rethrow(e);
+                                } finally {
+                                    if (handleInputStream != null) {
+                                        IOUtils.closeQuietly(handleInputStream);
+                                    }
+                                }
+
+                                return fileStream;
+                            });
+
+            FileInputStream fin = null;
+            synchronized (cache) {
```

Comment:
Could you explain why do we need `synhronized` here at all?

Reply:
Hi @rkhachatryan , synchronized here make sure no other thread decrease the ref count and delete the cache file.

Reply:
I think that should be achieved by reference counting (which is backed by an atomic integer).